lib/linked_list.rb

```diff
@@ -7,39 +7,71 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if head.nil?
+    return @head.data if !head.nil?
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(n)
   def length
-    return 0
+    if @head.nil?
+      return 0
+    end
+    current = @head
+    length = 1
+    until current.next.nil?
+      length += 1
+      current = current.next
+    end
+    return length
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(n)
   def add_last(data)
-
+    if @head.nil?
+      @head = Node.new(data)
+    else
+      last = self.find_last
+      last.next = Node.new(data)
+    end
+    
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
-  def get_last
+  def find_last
+    current = @head
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if head.nil?
    current = head
```